File: schemas/crm/drizzle/deal_activities.ts
// deal_activities: append-only audit trail of deal changes for pipeline analytics.
// See README.md for full design rationale.

import { pgTable, pgEnum, uuid, text, timestamp, index } from "drizzle-orm/pg-core";
import { deals } from "./deals";
import { users } from "./users";

export const dealActivityActionEnum = pgEnum("deal_activity_action", [
  "created",
  "updated",
  "stage_changed",
  "won",
  "lost",
  "reopened",
]);

export const dealActivities = pgTable(
  "deal_activities",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    dealId: uuid("deal_id").notNull().references(() => deals.id, { onDelete: "cascade" }),
    userId: uuid("user_id").references(() => users.id, { onDelete: "set null" }),
    action: dealActivityActionEnum("action").notNull(),
    field: text("field"),
    oldValue: text("old_value"),
    newValue: text("new_value"),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index("idx_deal_activities_deal_id_created_at").on(table.dealId, table.createdAt),
    index("idx_deal_activities_user_id").on(table.userId),
  ]
);

File: schemas/crm/spacetimedb/deal_activities.rs
// deal_activities: Append-only audit trail of deal changes for pipeline analytics.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum DealActivityAction {
    Created, // type: String
    Updated,
    StageChanged,
    Won,
    Lost,
    Reopened,
}

#[spacetimedb::table(name = deal_activities, public)]
pub struct DealActivity {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub deal_id: String, // UUID — FK → deals.id (cascade delete)
    #[index(btree)]
    pub user_id: Option<String>, // UUID — FK → users.id (set null)
    pub action: DealActivityAction,
    pub field: Option<String>,
    pub old_value: Option<String>,
    pub new_value: Option<String>,
    pub created_at: Timestamp,
}
